storage/client/bucket_metadata.cc

```diff
@@ -0,0 +1,84 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "storage/client/bucket_metadata.h"
+#include "storage/client/internal/nljson.h"
+#include "storage/client/internal/parse_rfc3339.h"
+
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+BucketMetadata BucketMetadata::ParseFromJson(std::string const& payload) {
+  auto json = storage::internal::nl::json::parse(payload);
+  BucketMetadata result{};
+  result.etag_ = json["etag"].get<std::string>();
+  result.id_ = json["id"].get<std::string>();
+  result.kind_ = json["kind"].get<std::string>();
+  if (json.count("labels") > 0) {
+    auto labels = json["labels"];
+    for (auto kv = labels.begin(); kv != labels.end(); ++kv) {
+      result.labels_.emplace(kv.key(), kv.value().get<std::string>());
+    }
+  }
+  result.location_ = json["location"].get<std::string>();
+  result.metadata_generation_ =
+      std::stoll(json["metageneration"].get_ref<std::string const&>());
+  result.name_ = json["name"].get<std::string>();
+  result.project_number_ =
+      std::stoll(json["projectNumber"].get_ref<std::string const&>());
+  result.self_link_ = json["selfLink"].get<std::string>();
+  result.storage_class_ = json["storageClass"];
+  result.time_created_ = storage::internal::ParseRfc3339(json["timeCreated"]);
+  result.time_updated_ = storage::internal::ParseRfc3339(json["updated"]);
+  return result;
+}
+
+bool BucketMetadata::operator==(BucketMetadata const& rhs) const {
+  return etag_ == rhs.etag_ and id_ == rhs.id_ and kind_ == rhs.kind_ and
+         labels_ == rhs.labels_ and location_ == rhs.location_ and
+         metadata_generation_ == rhs.metadata_generation_ and
+         name_ == rhs.name_ and project_number_ == rhs.project_number_ and
+         self_link_ == rhs.self_link_ and
+         storage_class_ == rhs.storage_class_ and
+         time_created_ == rhs.time_created_ and
+         time_updated_ == rhs.time_updated_;
+}
+
+std::ostream& operator<<(std::ostream& os, BucketMetadata const& rhs) {
+  // TODO() - convert back to JSON for a nicer format.
```

Comment:
Should the TODO have a username or an issue number? Just a thought.

Reply:
Should be, fixed.